gui/integrity-dialog.cpp

```diff
@@ -0,0 +1,425 @@
+	_percentLabel->setLabel(Common::String::format("%u %%", progress));
+	_progressBar->setValue(progress);
+}
+
+void IntegrityDialog::setError(Common::U32String &msg) {
+	_errorText->setLabel(msg);
+
+	_cancelButton->setLabel(_("Close"));
+	_cancelButton->setCmd(kCleanupCmd);
+}
+
+void IntegrityDialog::calculateTotalSize(Common::String gamePath) {
+	const Common::FSNode dir(gamePath);
+
+	if (!dir.exists() || !dir.isDirectory())
+		return;
+
+	Common::FSList fileList;
+	if (!dir.getChildren(fileList, Common::FSNode::kListAll))
+		return;
+
+	if (fileList.empty())
+		return;
+
+	// Process the files and subdirectories in the current directory recursively
+	for (Common::FSList::const_iterator it = fileList.begin(); it != fileList.end(); it++) {
+		const Common::FSNode &entry = *it;
+
+		if (entry.isDirectory())
+			calculateTotalSize(entry.getPath());
+		else {
+			Common::File file;
+			if (!file.open(entry))
+				continue;
+
+			g_state->totalSize += file.size();
+		}
+	}
+}
+
+Common::Array<Common::StringArray> IntegrityDialog::generateChecksums(Common::String gamePath, Common::Array<Common::StringArray> &fileChecksums) {
+	const Common::FSNode dir(gamePath);
+
+	if (!dir.exists() || !dir.isDirectory())
+		return {};
+
+	Common::FSList fileList;
+	if (!dir.getChildren(fileList, Common::FSNode::kListAll))
+		return {};
+
+	if (fileList.empty())
+		return {};
+
+	// Process the files and subdirectories in the current directory recursively
+	for (Common::FSList::const_iterator it = fileList.begin(); it != fileList.end(); it++) {
+		const Common::FSNode &entry = *it;
+
+		if (entry.isDirectory())
+			generateChecksums(entry.getPath(), fileChecksums);
+		else {
+			const Common::Path filename(entry.getPath());
+			auto macFile = Common::MacResManager();
+			if (macFile.openFileOrDataFork(filename)) {
+				debug("got here lol");
+				auto originalStream = macFile.openFileOrDataFork(filename);
+				auto fileStream = originalStream;
+
+				Common::Array<Common::String> fileChecksum = {filename.toString()};
+				// Various checksizes
+				for (auto size : {0, 5000, 1024 * 1024}) {
+					fileChecksum.push_back(Common::computeStreamMD5AsString(*(fileStream), size).c_str());
+					fileStream->seek(0);
+				}
+				// Tail checksums with checksize 5000
+				fileStream->seek(-5000, SEEK_END);
+				fileChecksum.push_back(Common::computeStreamMD5AsString(*(fileStream)).c_str());
+
+				g_state->calculatedSize += fileStream->size();
+
+				fileChecksums.push_back(fileChecksum);
+			}
+
+			Common::File file;
+			if (!file.open(entry))
+				continue;
+
+			Common::Array<Common::String> fileChecksum = {filename.toString()};
+			// Various checksizes
+			for (auto size : {0, 5000, 1024 * 1024}) {
+				fileChecksum.push_back(Common::computeStreamMD5AsString(file, size).c_str());
+				file.seek(0);
+			}
+			// Tail checksums with checksize 5000
+			file.seek(-5000, SEEK_END);
+			fileChecksum.push_back(Common::computeStreamMD5AsString(file).c_str());
```

Comment:
I wonder, how do you distinguish between all of these checksums in the array? I see you pushing just the checksums.